File: src/Java/36.IsValidSudoku.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    public boolean isValidSudoku(char[][] board) {
        boolean flag=true;
        for(int i=0;i<9;i++){
            int[] a = new int[]{0,0,0,0,0,0,0,0,0,0};
            for(int j=0;j<9;j++){
                if(board[i][j]!='.'){
                    if(a[board[i][j]-'0']==0)
                        a[board[i][j]-'0']=1;
                    else
                        flag=false; 
                }
            }
            if(flag==false)
               return flag;
        }   
        for(int i=0;i<9;i++){
            int[] a = new int[]{0,0,0,0,0,0,0,0,0,0};
            for(int j=0;j<9;j++){
                if(board[j][i]!='.'){
                    if(a[board[j][i]-'0']==0)
                        a[board[j][i]-'0']=1;
                    else
                        flag=false; 
                }
            }
            if(flag==false)
               return flag;
        }     
        for(int i=0;i<7;i+=3){
            for(int j=0;j<7;j+=3){
                int[] a = new int[]{0,0,0,0,0,0,0,0,0,0};
                for(int p=i;p<i+3;p++){
                    for(int q=j;q<j+3;q++){
                        if(board[p][q]!='.'){
                    if(a[board[p][q]-'0']==0)
                        a[board[p][q]-'0']=1;
                    else
                        flag=false; 
                }
                    }
                }
            if(flag==false)
               return flag;
            }
        }      
        return flag; 
    }
}

File: src/CSharp/36.IsValidSudoku.cs
using System;
using System.Collections.Generic;

namespace PeefyLeetCode.IsValidSudoku
{
public class Solution {
    public bool IsValidSudoku(char[,] board) {
        for (int i = 0; i < 9; i++) {
            int[] row = new int[9];
            int[] col = new int[9];
            int[] cube = new int[9];
            for (int j = 0; j < 9; j++) {
                if (board[i,j] != '.') {
                    if (row[board[i,j] - '1'] == 1) {
                        return false;
                    } else {
                        row[board[i,j] - '1'] = 1;
                    }
                }

                if (board[j,i] != '.') {
                    if (col[board[j,i] - '1'] == 1) {
                        return false;
                    } else {
                        col[board[j,i] - '1'] = 1;
                    }
                }
                // 每一宫内行列的变化
                int cubeX = 3 * (i / 3) + j / 3;
                int cubeY = 3 * (i % 3) + j % 3;
                if (board[cubeX,cubeY] != '.') {
                    if (cube[board[cubeX,cubeY] - '1'] == 1) {
                        return false;
                    } else {
                        cube[board[cubeX,cubeY] - '1'] = 1;
                    }
                }
            }
        }
        return true;
    }
}
}